src/MongoDB.Bson/ObjectModel/BsonBinaryData.cs

```diff
@@ -105,28 +117,41 @@ public BsonBinaryData(byte[] bytes, BsonBinarySubType subType, GuidRepresentatio
                     throw new ArgumentException(message, nameof(guidRepresentation));
                 }
             }
-            _bytes = bytes;
-            _subType = subType;
+
             _guidRepresentation = guidRepresentation;
         }
 
         /// <summary>
         /// Initializes a new instance of the BsonBinaryData class.
         /// </summary>
         /// <param name="guid">A Guid.</param>
+#pragma warning disable 618
+        [Obsolete("Use the constructor that also takes a GuidRepresentation instead.")]
         public BsonBinaryData(Guid guid)
-            : this(guid, BsonDefaults.GuidRepresentation)
+            : this(
+                guid, 
+                BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2
+                    ? BsonDefaults.GuidRepresentation
+                    : throw new InvalidOperationException("This constructor can only be used when BsonDefaults.GuidRepresentationMode is V2."))
         {
         }
+#pragma warning restore 618
 
         /// <summary>
         /// Initializes a new instance of the BsonBinaryData class.
         /// </summary>
         /// <param name="guid">A Guid.</param>
         /// <param name="guidRepresentation">The representation for Guids.</param>
         public BsonBinaryData(Guid guid, GuidRepresentation guidRepresentation)
-            : this(GuidConverter.ToBytes(guid, guidRepresentation), (guidRepresentation == GuidRepresentation.Standard) ? BsonBinarySubType.UuidStandard : BsonBinarySubType.UuidLegacy, guidRepresentation)
         {
+            _bytes = GuidConverter.ToBytes(guid, guidRepresentation);
+            _subType = GuidConverter.GetSubType(guidRepresentation);
+#pragma warning disable 618
+            if (BsonDefaults.GuidRepresentationMode == GuidRepresentationMode.V2)
```

Comment:
I don't really like a lot of conditions like this.. We can have more good looking code if we add an extension method to the GuidRepresentation enum like: `TrySetGuidRepresentation` and use this method everywhere where we're trying to set this value, but I remember we were not comfortable in using extensions for enums, so I'm ok with it